structs-in-rust/article.md

```diff
@@ -0,0 +1,443 @@
+- It must be implemented manually.
+- It is expected to format values in a more pretty way, without containing any unnecessary information.
+
+You can think of the `Debug` and the `Display` as the formatters for programmers and users respectively.
+
+Let's add the manual implementation of the `Display`:
+
+```
+impl std::fmt::Display for Point {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "({}, {})", self.x, self.y)
+    }
+}
+
+let p = Point::new(0, 1);
+
+// format a pair
+assert_eq!(format!("{:?}", p), String::from("Point { x: 0, y: 0 }"));
+assert_eq!(format!("{}", p), String::from("(0, 0)"));
+```
+
+The syntax seems pretty natural. We `impl`ement the trait `std::fmt::Display` `for` the struct called `Point`. In the opened code block we then have to define every function that is declared in the trait. In our case, there are only one — `fmt`.
+
+## Struct methods
+
+Just like in Java or C++, in Rust, you can define **methods** for a particular type. This is done in an `impl` block.
+
+```
+impl Point {
+    // Associated function
+    fn new(x: i32, y: i32) -> Point {
+        Point { x, y }
+    }
+
+    // Immutable method
+    fn has_same_x(&self, other: &Self) -> bool {
+        self.x == other.x
+    }
+
+    // Mutable method
+    fn shift_right(&mut self, dx: i32) {
+        self.x += dx;
+    }
+}
+
+// Mutable binding
+let mut point = Point::new(25, 25);
+assert!(!point.has_same_x(&Point::new(30, 25)));
+
+// Calling a method that changes the object state
+point.shift_right(3);
+assert_eq!(point.x, 28);
+```
+
+There are two types of functions related to a particular type.
+
+- Associated functions. They don't require an instance to be called on. Like static functions in Java.
+- Methods. They are invoked on an instance of a type.
+
+### Self
+
+You may have noticed that methods have `self` as their first argument. It represents the instance of the struct the method is being called on, similar to how it is done in Python. Some syntax sugar is involved here. Let's desugar it in two steps.
+
+|     |                                              |                                             |
+| --- | -------------------------------------------- | ------------------------------------------- |
+| 1   | `fn has_same_x(&self, other: &Self)`         | `fn shift_right(&mut self, dx: i32)`        |
+| 2   | `fn has_same_x(self: &Self, other: &Self)`   | `fn shift_right(self: &mut Self, dx: i32)`  |
+| 3   | `fn has_same_x(self: &Point, other: &Point)` | `fn shift_right(self: &mut Point, dx: i32)` |
+
+
+`Self` is an alias for the type of the impl block. In our case, it is `Point`.
+
+Methods can mutate the struct instance they are associated with, but this requires `&mut self` as the first argument. This way they can be called only via mutable binding.
+
+### Why use methods instead of regular functions?
+
+There are at least two reasons.
```

Comment:
Another tiny advantage of methods is autoreferencing:
You can write
```
point.shift_right(3);
```
instead of explicit
```
(&mut point).shift_right(3);
```

For functions, you always have to reference:
```
shift_point_right(&mut point, 3);
```

Reply:
Added